add docs for mxcube-web/server.yaml config file (#1190)
mxcube/mxcube3: docs/source/config/server.rst

server.yaml
===========

> The ``server.yaml`` file is used to configure various aspects of the MxCube's back-end server.

Reviewer: Casing of MXCuBE

Author: changed to MXCuBE

The configuration file contains the following sections:

* server_
* mxcube_

See below an `example  server.yaml <server_yaml_example_>`_ file.

server
------

> The server section primarily deals with the configuration of the HTTP aspects of the back-end server.

Reviewer 2: ```suggestion
The ``server`` section primarily deals with the configuration of the HTTP aspects of the back-end server.
```

> The ``server`` section primarily deals with the configuration of the HTTP aspects of the back-end server.

Reviewer: I would perhaps omit the part about "HTTP aspects" and replace it with maybe web server aspects or similar

Author: Yes, it's is strictly speaking it is not just HTTP.  I have changed to 'network aspects'.

It allows customizing the behavior of the Flask framework, its plugins and other components.
The section has the following syntax:

.. code-block:: yaml

    server:
      DEBUG: <boolean>
      SECRET_KEY: <secret key>
      SECURITY_PASSWORD_SALT: <password salt>
      SECURITY_TRACKABLE: <boolean>
      PERMANENT_SESSION_LIFETIME: <session lifetime>
      USER_DB_PATH: <db_file_path>
      CERT_KEY: <certificate key file path>
      CERT_PEM: <certificate pem file path>
      CERT: <ssl_mode>
      ALLOWED_CORS_ORIGINS:
        - <allowed URL>
        - <allowed URL>
        ...

The ``server`` section may contain keys as described below.

``DEBUG``
~~~~~~~~~

Whether Flask debug mode is enabled.

> This is equivalent of Flask's `DEBUG <https://flask.palletsprojects.com/en/2.3.x/config/#DEBUG>`_ configuration value.

Reviewer 2: ```suggestion
This is the equivalent of Flask's `DEBUG <https://flask.palletsprojects.com/en/2.3.x/config/#DEBUG>`_ configuration value.
```

Default value is ``/tmp/mxcube-user.db``.

``ALLOWED_CORS_ORIGINS``
~~~~~~~~~~~~~~~~~~~~~~~~

List of origins that are allowed to connect to this server.
This list is passed via ``cors_allowed_origins`` parameter to the `SocketIO <https://flask-socketio.readthedocs.io/en/latest/api.html#flask_socketio.SocketIO>`_ class constructor.
Note that if this list is too restrictive, the front-end will fail to create a SocketIO connection to the back-end.

``CERT``
~~~~~~~~

Used to enable SSL and set the type of host certificate.
The following values are supported:

+--------+------------------------------------------------+
| NONE   | SSL disabled                                   |
+--------+------------------------------------------------+
| ADHOC  | SSL enabled, autogenerated certificate is used |
+--------+------------------------------------------------+
| SIGNED | SSL enabled, signed certificate is used        |
+--------+------------------------------------------------+

Default value is ``NONE``.

``CERT_PEM``
~~~~~~~~~~~~

Path of the certificate file to use for SSL connections.

``CERT_KEY``
~~~~~~~~~~~~

Path of the private key file to use for SSL connections.

mxcube
------

> The mxcube section allows you to customize miscellaneous aspects of the MxCube application.
> It have settings for the on-axis video stream, user account management, experiment mode.

Reviewer 2: ```suggestion
The ``mxcube`` section allows you to customize miscellaneous aspects of the MXCuBE application.
It contains settings for the on-axis video stream, user account management, and experiment mode.
```

The section has the following syntax:

.. code-block:: yaml

    mxcube:
      mode: <experiment mode>
      USE_EXTERNAL_STREAMER: <boolean>
      VIDEO_FORMAT: <format name>
      VIDEO_STREAM_URL: <stream URL>
      VIDEO_STREAM_PORT: <stream port>
      usermanager:
        class: <class name>
        inhouse_is_staff: <boolean>
        users:
          - username: <user name>
            role: <role name>
          ...

The ``mxcube`` section may contain keys as described below.

> ``mode``

Reviewer 2: Unrelated to the documentation, but wondering if this should be changed to ~`PASCAL_CASE`~`CONSTANT_CASE` for consistency (same with other `snake_case` keys).

Author: You mean the mode names? Like:

    SSX-CHIP -> SSX_CHIP
    SSX-INJECTOR -> SSX_INJECTOR

yes, I agree.

Reviewer 2: Sorry I meant the key itself, `MODE`, for consistency with the other keys.

... unless I've missed that the distinction matters for some reason? Are ~`PASCAL_CASE`~`CONSTANT_CASE` keys ones that can be overridden via environment variables or something? I don't necessarily think that keys in a YAML file should be written in ~`PASCAL_CASE`~`CONSTANT_CASE` (unlike keys in a `.env` file for instance), but it should definitely be consistent (or the reason why it's not should perhaps be documented).

Reviewer 3: _[Aside: @axelboc, this is [`PascalCase`](https://en.wikipedia.org/wiki/Camel_case), isn't it? You probably mean [`SCREAMING_SNAKE_CASE`](https://en.wikipedia.org/wiki/Snake_case), no?]_

Reviewer 2: Haha yeah of course, sorry ^^ -- I actually use the term `CONSTANT_CASE` usually, sorry for the mix-up...

Reviewer 3: _[@axelboc, the names are silly, and I always have to look them up anyway. Seeing `CONSTANT_CASE` gets the message across much better for me, indeed.]_

Author: I agree that we should make the casing consistent in server.yaml.

But think that's a breaking change, so needs to be carefully managed. In other words this is a topic for another PR :)

~~~~~~~~

Set the experiment mode for the application.
Different modes tailor the presented UI to be suitable for a specific type of MX experiments.
The following values are supported:

+--------------+--------------------------------------+
| mode         | experiment type                      |
+==============+======================================+
| OSC          | standard oscillation data collection |
+--------------+--------------------------------------+
| SSX-CHIP     | SSX fixed target experiments         |
+--------------+--------------------------------------+
| SSX-INJECTOR | SSX injector experiments             |
+--------------+--------------------------------------+

Default mode is ``OSC``.

``USE_EXTERNAL_STREAMER``
~~~~~~~~~~~~~~~~~~~~~~~~~

Use an external process to stream the sample view video.

> When enabled, MxCube will invoke an appropriate hook on the camera hardware object to start the streamer.

Reviewer 2: ```suggestion
When enabled, MXCuBE will invoke an appropriate hook on the camera hardware object to start the streamer.
```

Reviewer 3: Fun fact: there is a substitution `|project|` pre-configured in the `rst_prolog`:

https://github.com/mxcube/mxcubeweb/blob/ec1048b3d8fbf9360e6000a3b1f6264f26fa390f/docs/conf.py#L42-L44

So in principle we could write this:
```suggestion
When enabled, |project| will invoke an appropriate hook on the camera hardware object to start the streamer.
```
... and it should show up as `MXCuBE-Web` in the resulting output (HTML), if I did not mess things up.

Now I am think it might make sense to add it as `|mxcube|` in stead of (or in addition to) `|project|`. 

I have not really investigated if/how we could do this for Markdown documents.

Also maybe I picked the wrong example, and in this case it is really "MXCuBE" that is meant, and not "MXCuBE-Web", but the point still stands.

It is assumed that the camera object supports the external streamer feature.

Default value is ``False``.

``VIDEO_FORMAT``
~~~~~~~~~~~~~~~~

The format to use when streaming the sample view video.
The ``MPEG1`` and ``MJPEG`` formats are supported.
Note that your camera hardware object must support the specified format.

Default format is ``MPEG1``.

``VIDEO_STREAM_URL``
~~~~~~~~~~~~~~~~~~~~

> When set, it will override the default video stream URL.
> The URL that is used by the UI to fetch the sample view video.

Reviewer 2: I would swap the two sentences:

```suggestion
The URL that is used by the UI to fetch the sample view video.
When set, it will override the default video stream URL.
```

This configuration is primarily useful with external streams,
that provide video stream from custom URLs.

``VIDEO_STREAM_PORT``
~~~~~~~~~~~~~~~~~~~~~

This configuration is used when ``USE_EXTERNAL_STREAMER`` flag is enabled.
The specified value is passed on to the camera hardware object,
when invoking the start stream hook.

``usermanager``
~~~~~~~~~~~~~~~

This subsection allows configuring user management in the application.

> ``class`` key specifies the class name of the user manager.

Reviewer 2: For consistency:

```suggestion
``class`` specifies the class name of the user manager.
```

> This allows you use a custom implementation of the user manager if needed.

Reviewer 2: ```suggestion
This allows using a custom implementation of the user manager if needed.
```

The default class name is ``UserManager``.

``inhouse_is_staff`` specifies if all in-house users should automatically acquire *staff* role.
Which users are considered in-house is managed by the session hardware object.

The default value is ``True``.

> ``users`` key allows to configure users that should acquire additional roles.

Reviewer 2: ```suggestion
``users`` specifies users that should acquire additional roles.
```

> The contents of this key are a list of username and role specifications.

Reviewer 2: Might be worth adding an example too.

```suggestion
The content of this key is a list of username and role specifications.
```

> Currently only *staff* role can be specified.

Reviewer 2: ```suggestion
Currently, only the *staff* role can be specified.
```
